oauthenticator/oauth2.py

```diff
@@ -426,8 +481,274 @@ def get_handlers(self, app):
+        be extracted this way and needs extra processing.
+
+        Args:
+            user_info: the dictionary returned by the userdata request
+
+        Returns:
+            user_info["self.username_claim"] or raises an error if such value isn't found.
+
+        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`
+        """
+        username = user_info.get(self.username_claim, None)
+        if not username:
+            message = "No %s found in %s", self.username_claim, user_info
+            self.log.error(message)
+            raise ValueError(message)
+
+        return username
+
+    # Originally a GoogleOAuthenticator only feature
+    async def get_prev_refresh_token(self, handler, username):
+        """
+        Retrieves the `refresh_token` from previous encrypted auth state.
+        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`
+        """
+        user = handler.find_user(username)
+        if not user or not user.encrypted_auth_state:
+            return
+
+        self.log.debug(
+            "Encrypted_auth_state was found, will try to decrypt and pull refresh_token from it..."
+        )
+
+        try:
+            encrypted = user.encrypted_auth_state
+            auth_state = await decrypt(encrypted)
+
+            return auth_state.get('refresh_token')
+        except (ValueError, InvalidToken, EncryptionUnavailable) as e:
+            self.log.warning(
+                f"Failed to retrieve encrypted auth_state for {username}. Error was {e}.",
+            )
+            return
+
+    def build_access_tokens_req_params(self, handler):
+        """
+        Builds the parameters that should be passed to the URL request
+        that exchanges the OAuth code for the Access Token.
+        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`.
+        """
+        code = handler.get_argument("code")
+        if not code:
+            raise web.HTTPError(400, "Authentication Cancelled.")
+
+        params = {
+            'code': code,
+            'grant_type': 'authorization_code',
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'redirect_uri': self.get_callback_url(handler),
+        }
+        params.update(self.extra_params)
+
+        return params
+
+    async def get_tokens_info(self, handler, params):
+        """
+        Makes a "POST" request to `self.token_url`, with the parameters received as argument.
+
+        Returns:
+            the JSON response to the `token_url` the request.
+
+        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`
+        """
+        url = url_concat(self.token_url, params)
+
+        req = HTTPRequest(
+            url,
+            method="POST",
+            headers=self.build_token_info_req_headers(),
+            body=json.dumps(params),
+            validate_cert=self.validate_server_cert,
+        )
+
+        tokens_info = await self.fetch(req)
```

Comment:
Consistency: let's use `token_info` everywhere, instead of plural `tokens_info`.